File: apps/api/momento_api/routes/share.py
import secrets
from datetime import datetime, timedelta, timezone
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status

from momento_api.auth.dependencies import CurrentUser, get_current_user
from momento_api.auth.password import hash_password
from momento_api.database import execute_query, fetch_all, fetch_one, get_connection, insert_returning_id
from momento_api.models.share import ShareCreateRequest, ShareDeleteRequest, ShareLinkResponse, ShareListResponse

router = APIRouter(prefix="/share", tags=["share"])


def _row_to_share_response(row) -> ShareLinkResponse:
    return ShareLinkResponse(
        id=row["id"],
        token=row["token"],
        media_id=row["media_id"],
        album_id=row["album_id"],
        has_password=row["password_hash"] is not None,
        expires_at=row["expires_at"],
        view_count=row["view_count"],
        created_at=row["created_at"],
    )


@router.post("/create", response_model=ShareLinkResponse)
async def create_share_link(
    request: ShareCreateRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> ShareLinkResponse:
    if request.media_id is None and request.album_id is None:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Must specify media_id or album_id")

    if request.media_id is not None and request.album_id is not None:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Cannot specify both media_id and album_id")

    if request.media_id:
        media_row = fetch_one("SELECT id FROM media WHERE id = ? AND user_id = ?", (request.media_id, current_user.id))
        if media_row is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Media not found")

    if request.album_id:
        album_row = fetch_one("SELECT id FROM albums WHERE id = ? AND user_id = ?", (request.album_id, current_user.id))
        if album_row is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Album not found")

    token = secrets.token_urlsafe(16)
    password_hash = hash_password(request.password) if request.password else None
    expires_at = None
    if request.expires_in_days:
        expires_at = (datetime.now(timezone.utc) + timedelta(days=request.expires_in_days)).isoformat()

    share_id = insert_returning_id(
        """
        INSERT INTO share_links (user_id, media_id, album_id, token, password_hash, expires_at)
        VALUES (?, ?, ?, ?, ?, ?)
        """,
        (current_user.id, request.media_id, request.album_id, token, password_hash, expires_at),
    )

    row = fetch_one("SELECT * FROM share_links WHERE id = ?", (share_id,))
    if row is None:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Failed to create share link")

    return _row_to_share_response(row)


@router.post("/list", response_model=ShareListResponse)
async def list_share_links(current_user: Annotated[CurrentUser, Depends(get_current_user)]) -> ShareListResponse:
    rows = fetch_all("SELECT * FROM share_links WHERE user_id = ? ORDER BY created_at DESC", (current_user.id,))
    return ShareListResponse(shares=[_row_to_share_response(row) for row in rows])


@router.post("/delete")
async def delete_share_link(
    request: ShareDeleteRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> dict:
    share_row = fetch_one(
        "SELECT id FROM share_links WHERE id = ? AND user_id = ?", (request.share_id, current_user.id)
    )
    if share_row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Share link not found")

    execute_query("DELETE FROM share_links WHERE id = ?", (request.share_id,))
    get_connection().commit()

    return {"message": "Share link deleted successfully"}

File: apps/api/src/routes/share.rs
use axum::{extract::State, routing::post, Json, Router};
use chrono::{Duration, Utc};
use rand::Rng;

use crate::auth::{hash_password, AppState, CurrentUser};
use crate::database::{execute_query, fetch_all, fetch_one, insert_returning_id};
use crate::error::{AppError, AppResult};
use crate::models::{ShareCreateRequest, ShareDeleteRequest, ShareLinkResponse, ShareListResponse};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/share/create", post(create_share_link))
        .route("/share/list", post(list_share_links))
        .route("/share/delete", post(delete_share_link))
}

fn map_share_row(row: &rusqlite::Row) -> rusqlite::Result<ShareLinkResponse> {
    let password_hash: Option<String> = row.get(4)?;
    Ok(ShareLinkResponse {
        id: row.get(0)?,
        token: row.get(1)?,
        media_id: row.get(2)?,
        album_id: row.get(3)?,
        has_password: password_hash.is_some(),
        expires_at: row.get(5)?,
        view_count: row.get(6)?,
        created_at: row.get(7)?,
    })
}

async fn create_share_link(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<ShareCreateRequest>,
) -> AppResult<Json<ShareLinkResponse>> {
    if request.media_id.is_none() && request.album_id.is_none() {
        return Err(AppError::BadRequest(
            "Must specify media_id or album_id".to_string(),
        ));
    }

    if request.media_id.is_some() && request.album_id.is_some() {
        return Err(AppError::BadRequest(
            "Cannot specify both media_id and album_id".to_string(),
        ));
    }

    let conn = state.pool.get().map_err(AppError::Pool)?;

    if let Some(media_id) = request.media_id {
        let exists = fetch_one(
            &conn,
            "SELECT id FROM media WHERE id = ? AND user_id = ?",
            &[&media_id, &current_user.id],
            |row| Ok(row.get::<_, i64>(0)?),
        )?;

        if exists.is_none() {
            return Err(AppError::NotFound("Media not found".to_string()));
        }
    }

    if let Some(album_id) = request.album_id {
        let exists = fetch_one(
            &conn,
            "SELECT id FROM albums WHERE id = ? AND user_id = ?",
            &[&album_id, &current_user.id],
            |row| Ok(row.get::<_, i64>(0)?),
        )?;

        if exists.is_none() {
            return Err(AppError::NotFound("Album not found".to_string()));
        }
    }

    // Generate token
    let token: String = rand::thread_rng()
        .sample_iter(&rand::distributions::Alphanumeric)
        .take(22)
        .map(char::from)
        .collect();

    let password_hash = request
        .password
        .as_ref()
        .map(|p| hash_password(p))
        .transpose()
        .map_err(|e| AppError::Internal(format!("Failed to hash password: {}", e)))?;

    let expires_at = request
        .expires_in_days
        .map(|days| (Utc::now() + Duration::days(days as i64)).to_rfc3339());

    let share_id = insert_returning_id(
        &conn,
        "INSERT INTO share_links (user_id, media_id, album_id, token, password_hash, expires_at) VALUES (?, ?, ?, ?, ?, ?)",
        &[
            &current_user.id,
            &request.media_id,
            &request.album_id,
            &token,
            &password_hash,
            &expires_at,
        ],
    )?;

    let share = fetch_one(
        &conn,
        "SELECT id, token, media_id, album_id, password_hash, expires_at, view_count, created_at FROM share_links WHERE id = ?",
        &[&share_id],
        map_share_row,
    )?
    .ok_or_else(|| AppError::Internal("Failed to create share link".to_string()))?;

    Ok(Json(share))
}

async fn list_share_links(
    State(state): State<AppState>,
    current_user: CurrentUser,
) -> AppResult<Json<ShareListResponse>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let shares = fetch_all(
        &conn,
        "SELECT id, token, media_id, album_id, password_hash, expires_at, view_count, created_at FROM share_links WHERE user_id = ? ORDER BY created_at DESC",
        &[&current_user.id],
        map_share_row,
    )?;

    Ok(Json(ShareListResponse { shares }))
}

async fn delete_share_link(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<ShareDeleteRequest>,
) -> AppResult<Json<serde_json::Value>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let exists = fetch_one(
        &conn,
        "SELECT id FROM share_links WHERE id = ? AND user_id = ?",
        &[&request.share_id, &current_user.id],
        |row| Ok(row.get::<_, i64>(0)?),
    )?;

    if exists.is_none() {
        return Err(AppError::NotFound("Share link not found".to_string()));
    }

    execute_query(
        &conn,
        "DELETE FROM share_links WHERE id = ?",
        &[&request.share_id],
    )?;

    Ok(Json(serde_json::json!({"message": "Share link deleted successfully"})))
}
